presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewQueryOptimizer.java

```diff
@@ -96,17 +147,60 @@ protected Node visitQuerySpecification(QuerySpecification node, Void context)
-        // TODO: Handle filter containment problem https://github.com/prestodb/presto/issues/16405
-        if (materializedViewInfo.getWhereClause().isPresent() && !materializedViewInfo.getWhereClause().equals(node.getWhere())) {
-            throw new IllegalStateException("Query with no where clause is not rewritable by materialized view with where clause");
-        }
         if (materializedViewInfo.getGroupBy().isPresent() && !node.getGroupBy().isPresent()) {
             throw new IllegalStateException("Query with no groupBy clause is not rewritable by materialized view with groupBy clause");
         }
         // TODO: Add HAVING validation to the validator https://github.com/prestodb/presto/issues/16406
         if (node.getHaving().isPresent()) {
             throw new SemanticException(NOT_SUPPORTED, node, "Having clause is not supported in query optimizer");
         }
+        if (materializedViewInfo.getWhereClause().isPresent()) {
+            if (!node.getWhere().isPresent()) {
+                throw new IllegalStateException("Query with no where clause is not rewritable by materialized view with where clause");
+            }
+            if (!(node.getFrom().get() instanceof Table)) {
+                throw new SemanticException(NOT_SUPPORTED, node, "Relation other than Table is not supported in query optimizer");
+            }
+            Table baseTable = (Table) node.getFrom().get();
+            QualifiedObjectName baseTableName = createQualifiedObjectName(session, baseTable, baseTable.getName());
+
+            Optional<TableHandle> tableHandle = metadata.getTableHandle(session, baseTableName);
+            if (!tableHandle.isPresent()) {
+                throw new SemanticException(MISSING_TABLE, node, "Table does not exist");
+            }
+
+            ImmutableList.Builder<Field> fields = ImmutableList.builder();
+
+            for (ColumnHandle columnHandle : metadata.getColumnHandles(session, tableHandle.get()).values()) {
+                ColumnMetadata columnMetadata = metadata.getColumnMetadata(session, tableHandle.get(), columnHandle);
+                fields.add(Field.newUnqualified(columnMetadata.getName(), columnMetadata.getType()));
+            }
+
+            Scope scope = Scope.builder()
+                    .withRelationType(RelationId.anonymous(), new RelationType(fields.build()))
+                    .build();
+
+            /**
+             * Given base query's filter condition and materialized view's filter condition, the goal is to check if MV's filters contain Base's filters (Base implies MV).
+             * Let base query's filter condition be A, and MV's filter condition be B.
+             * One way to evaluate A implies B is to evaluate logical expression A^~B and check if the output domain is none.
+             * If the resulting domain is none, then A^~B is false. Thus A implies B.
+             * For more information and examples: https://fb.quip.com/WwmxA40jLMxR
+             */
```

Comment:
- no need to add a java doc; comments are good enough
- can we also link the SAT solver ticket here?